### DISCUSSION TITLE
【已完成】组件定义 API 改造进度跟踪

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#3041] [分享 / Share]
### 改造目标

对组件的 options 提供完整的类型提示

### 改造方案

基本思想来自 discussion #3011 中的方案 1，其原始设计代码如下：

<details>
<summary>设计方案代码</summary>

```typescript
module component1 {
    import D = componentCreate

    // 定义组件
    export const component = D.createComponentMetadata({
        name: 'component1Name',
        options: {
            component1Data: {
                defaultValue: 0,
            },
        },
        // entry 的参数可以获得类型提示
        // 该类型来自 options 定义中的 defaultValue
        entry: (o) => o,
    })
}

module component2 {
    import D = componentCreate

    // 先定义 options
    const options = D.createOptionsMetadata({
        component2Data: {
            defaultValue: 0,
        },
    })

    // 自动推断 Options 的类型
    type Component2Options = D.OptionsOfMeta<typeof options>

    // entry 此时可在外部定义，且有类型提示
    const entry = (o: Component2Options) => o

    export const component = D.createComponentMetadata({
        name: 'component2Name',
        options,
        entry,
    })
}

module componentCreate {
    import T = componentType

    export function createOptionsMetadata<O extends T.UnknownOptions>(
        v: T.OptionsMetadata<O>
    ): T.OptionsMetadata<O> {
        return v
    }

    export function createComponentMetadata<O extends T.UnknownOptions>(
        v: T.ComponentMetadata<O>
    ): T.ComponentMetadata<O> {
        return v
    }

    export type OptionsOfMeta<M extends T.OptionsMetadata<any>> = (
        M extends T.OptionsMetadata<infer O> ? O : never
    )
}

module componentType {
    export interface UnknownOptions {
        [name: string]: unknown
    }

    export interface OptionMetadata<V = unknown> {
        defaultValue: V
    }

    export type OptionsMetadata<O extends UnknownOptions = UnknownOptions> = {
        [Name in keyof O]: OptionMetadata<O[Name]>
    }

    export interface ComponentMetadata<
        O extends UnknownOptions = UnknownOptions
    > {
        name: string
        options: OptionsMetadata<O>
        entry: (options: O) => void
    }

    export interface ComponentSettings<
        O extends UnknownOptions = UnknownOptions
    > {
        enabled: boolean
        options: O
    }

    // 获取组件设置，但使用自符串时无法推断类型，需要强转
    export function getComponentSettings<
        O extends UnknownOptions,
        C extends (ComponentMetadata<O> | string)
    >(component: C): (
        C extends string 
            ? ComponentSettings<UnknownOptions>
            : ComponentSettings<O>
    ) {
        // 省略实现
        return null as unknown as ComponentSettings<O>
    }
}
```

</details>

### 当前进度

- [x] 核心部分(PR #3143)：
  - [x] `src/components` 中的 `types.ts`, `define.ts`，以及 `src/core/settings` 中的 `helpers.ts`, `index.ts`
  - [x] 在 `src/components/type.ts` 中修改新方案，使其与旧代码兼容，并用 TODO 注释标出。
  - [x] 修改代码贡献指南——`CONTRIBUTING.md`。提示新组件需要使用新的定义方式。
- [x] 将旧组件改为新 api
  - [x] 内置组件 `src/components`（PR #3154）
  - [x] 动态相关组件 `registry/lib/components/feeds`（PR #3697）
  - [x] 直播相关组件 `registry/lib/components/live`（PR #3452）
  - [x] 样式相关组件 `registry/lib/components/style`（PR #3547）
  - [x] 触摸相关组件 `registry/lib/components/touch`（PR #3682）
  - [x] 工具相关组件 `registry/lib/components/util`（PR #3682）
  - [x] 视频相关组件 `registry/lib/components/video`（PR #3682）
  - [x] 弃用组件 `registry/lib/deprecated`（PR #3682）
  - [x] （如果方便的话）对二次封装的组件构造函数、类进行类型改造。如 switch options api 等（PR #3715）
- [x] 将 `src/components/types.ts` 中为兼容旧代码而修改的部分改回新方案代码（PR #3715）

### 废弃的 PR
#3038 #3141

**Comment by timongh** [comment]
前段时间比较忙，就没有忙这边。后面的时间会宽裕一些。

@the1812 
我发现 component-option-types 被合并到 preview-features 里面了。不知道是不是因为我太久没出现才这么做的，不过消失之前没打声招呼确实是我考虑不周，抱歉抱歉。

其实现在还不合适，因为这套类型里面还有一点问题，是我之前在修改时发现的。当时尝试过解决，只是还没有找到解决办法，可能解决不了也说不定。后面因为去忙自己的事了，就没继续。

这个问题具体来说就是，如果省略 defineComponentMetadata 里的 options 选项，那么其中 entry 函数推断出的参数是一个 UnknownOptions，而实际上此时这里应该是一个“空选项”（Record<string, never>）。

这个问题其实影响不大，我会继续尝试解决。不过 preview-features 分支该怎么办，可能还得你拿主意。

preview-features 那边，我本来的想法是，还按照原来的方式开发。component-option-types 这边同步跟上就行，因为这两种定义方式的区别并不大，要改过来其实并不算太麻烦。

**Reply by timongh** [reply]
解决方向我现在大概有一些，只是具体采用什么方式，怎么改，我还要再做做实验，考虑考虑。

我更希望的是，把 preview-features 回退到之前的定义方式。因为我更倾向的解决方案改动会比较大。

**Reply by the1812** [reply]
> 我发现 component-option-types 被合并到 preview-features 里面了

因为新功能的开发需要使用 define API, 并且不影响编译产物, 用的也是过渡方案 (Record<string, any>), 所以合并了也没关系的. 你可以将 component-option-types 同步一下 preview-features 并继续开发, 我会在需要 component-option-types 中的新功能时将其合并回 preview-features.

> 那么其中 entry 函数推断出的参数是一个 UnknownOptions，而实际上此时这里应该是一个“空选项”（Record<string, never>）

UnknownOptions (Record<string, unknown>) 才是对的, options 运行时可以被注入其他选项.

**Reply by timongh** [reply]
> options 运行时可以被注入其他选项.

我印象中没有注入options 的功能啊。什么时候会注入啊，哪些 API 会注入？有例子吗？

**Reply by the1812** [reply]
https://github.com/the1812/Bilibili-Evolved/blob/bc268c1fd2aecdce4fa4f610d48c3efa6dc14a8c/registry/lib/plugins/video/download/aria2-output/RpcConfig.vue#L142-L146

只要拿到 options 对象, 往上面写入新的值就可以了, 这类值不会在设置面板中展示

**Reply by timongh** [reply]
这种动态修改完全不是类型可以预料到的，因为用户可以拿着这个对象干任何事。除非全部用 unknown。所以这个地方还是用 Record<string, never> 更好。

如果真的要这么用，还是直接在用的时候强转吧。不过最好的方式，还是补一个声明上去。

**Reply by the1812** [reply]
也可以，那还是用 never 吧
